snips_nlu/intent_parser/lookup_intent_parser.py

```diff
@@ -0,0 +1,403 @@
+        """
+        if intent is None:
+            return []
+
+        if intent not in self.intents_names:
+            raise IntentNotFoundError(intent)
+
+        slots = self.parse(text, intents=[intent])[RES_SLOTS]
+        if slots is None:
+            slots = []
+        return slots
+
+    def get_intent_id(self, intent_name):
+        intent_id = self._intents_mapping.get(intent_name)
+        if intent_id is None:
+            intent_id = len(self.intents_names)
+            self.intents_names.append(intent_name)
+            self._slots_mapping[intent_name] = intent_id
+
+        return intent_id
+
+    def get_slot_id(self, slot_name):
+        slot_id = self._slots_mapping.get(slot_name)
+        if slot_id is None:
+            slot_id = len(self.slots_names)
+            self.slots_names.append(slot_name)
+            self._slots_mapping[slot_name] = slot_id
+
+        return slot_id
+
+    def _preprocess_text(self, txt):
+        """Replaces stop words and characters that are tokenized out by
+            whitespaces"""
+        tokens = tokenize(txt, self.language)
+        current_idx = 0
+        cleaned_string = ""
+        for token in tokens:
+            if self.stop_words and normalize_token(token) in self.stop_words:
+                token.value = "".join(" " for _ in range(len(token.value)))
+            prefix_length = token.start - current_idx
+            cleaned_string += "".join((" " for _ in range(prefix_length)))
+            cleaned_string += token.value
+            current_idx = token.end
+        suffix_length = len(txt) - current_idx
+        cleaned_string += "".join((" " for _ in range(suffix_length)))
+        return cleaned_string
+
+    def generate_io_mapping(self, intents, entity_placeholders):
+        """Generate input-output pairs"""
+        for intent_name, intent in iteritems(intents):
+            intent_id = self.get_intent_id(intent_name)
+            for entry in intent[UTTERANCES]:
+                yield self._build_io_mapping(intent_id, entry,
+                                             entity_placeholders)
+
+    def _build_io_mapping(self, intent_id, utterance, entity_placeholders):
+        input_ = []
+        output = [intent_id]
+        slots = []
+        for chunk in utterance[DATA]:
+            if SLOT_NAME in chunk:
+                slot_name = chunk[SLOT_NAME]
+                slot_id = self.get_slot_id(slot_name)
+                entity_name = chunk[ENTITY]
+                placeholder = entity_placeholders[entity_name]
+                input_.append(placeholder.lower())
+                slots.append(slot_id)
+            else:
+                input_.append(chunk[TEXT])
+        output.append(slots)
+
+        key = "".join(
+            [
+                t
+                for t in tokenize_light(" ".join(input_), self.language)
+                if normalize(t) not in self.stop_words
+            ]
+        )
+
+        return key.lower(), output
```

Comment:
This function (`_build_io_mapping`) should call `_preprocess_text` to avoid duplicating code, and to ensure consistency between what happens at train time and inference time.